### DISCUSSION TITLE
Imodels in multi viewports

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#3629] [Q&A] [answered]
Following [Four viewports](https://github.com/iTwin/itwinjs-core/discussions/3553#discussioncomment-2638610) we have four viewports displaying same imodel connection. We need to display different imodels in each of the viewports. How can we achieve the same?

Thanks.

**Comment by VedankNaik** [comment]
@pmconne you mentioned it could be done using APIs in @itwin/core-frontend [here](https://github.com/iTwin/itwinjs-core/discussions/3553#discussioncomment-2766369). How to go about it?

**Reply by pmconne** [reply]
Like I said in the reply you linked:

> I can tell you how to do either using the APIs in @itwin/core-frontend, that's easy.

Using just the APIs in @itwin/core-frontend, all you would need to do is:
- Define 4 `div`s in your document body.
- Open 4 `IModelConnection`s using `BriefcaseConnection.openFile`, `IModelConnection.openRemote`, or similar.
- Create a `ViewState` for each iModel, perhaps using `ViewCreator3d`.
- Pass a `div` and a `ViewState` to `ScreenViewport.create` for each of your 4 iModels.

> I suspect you are asking how to do it using the ui packages, which I cannot help with.

Based on your the snippet from your Four viewports link, you **are** using the ui packages. @NancyMcCallB and @bsteinbk may be able to provide guidance. I am not familiar with the ui layers.

> Please post a new, more specific question with relevant code etc.

Where is your code?

**Reply by VedankNaik** [reply]
For code I'm following Multiple Viewports sample with updates mentioned [here](https://github.com/iTwin/itwinjs-core/discussions/3553#discussioncomment-2638610)

**Reply by pmconne** [reply]
Yeah, that uses the Viewer component which tries to simplify things by handling many details for you, including opening the iModel. It assumes you only want to look at one iModel.
I'm certain there's a way to modify the sample to use two iModels, but I am not the right person to help with that.
I should point out that trying to synchronize two views of two different iModels will probably not work well, unless they are co-located and you're only synchronizing the frustum; or they are two versions of the same iModel in which element/category/model Ids refer to the same things in both iModels.

**Comment by bsteinbk** [comment, marked as answer]
You will also need to properly define the ContentGroup with all these connections and viewstates.  The example below does this statically or you could set up a ContentGroupProvider to serve up the ContentGroup definitions as in [this example](https://www.itwinjs.org/learning/ui/augmentingui/#adding-a-frontstage).


    // Create the content group.
    this._contentGroup = new ContentGroup({
      id: "MultiViewportContentGroup",
      layout: StandardContentLayouts.fourQuadrants,
      contents: [
        {
          id: "MultiViewport1",
          classId: IModelViewportControl,
          applicationData: {
            viewState1,
            iModelConnection: connection1,
          },
        },
        {
          id: "MultiViewport2",
          classId: IModelViewportControl,
          applicationData: {
            viewState2,
            iModelConnection: connection2,
          },
        },
        {
          id: "MultiViewport3",
          classId: IModelViewportControl,
          applicationData: {
            viewState3,
            iModelConnection: connection3,
          },
        },        {
          id: "MultiViewport4",
          classId: IModelViewportControl,
          applicationData: {
            viewState4,
            iModelConnection: connection4,
          },
        },
      ],
    });

**Reply by VedankNaik** [reply]
I tried opening multiple imodel connections and creating ViewStates from those using ViewCreator3d. I can see ViewStates being created. I used them inside ContentGroup but viewports appears blank. Used above snippet to do it statically. 
Do I need to make any more changes?

**Reply by pmconne** [reply]
@VedankNaik we are limited in our ability to debug your code without seeing/running your code. Perform some experiments. For example, keep your code the same except use the same iModel and ViewState for all viewports - do they still show blank? Check if all the models and/or categories are turned off, or turn them all on. Still blank? Try fitting the view. Check browser console for errors. Etc.

**Comment by VedankNaik** [comment]
Hello, 
Following above example on ContentGroup I was able to get four views. How do I make this dynamic? Consider I have button, how do I switch between 1 view and 4 views on button click?

**Comment by tonygem2022** [comment]
It seems the Baytown Cross Probing example is hardcoded as 2 screens. I see your issue. https://www.itwinjs.org/sandboxes/iTwinPlatform/Cross-Probing

Actually this would be an interesting functionality to have. How can we emulate this and how can we have the ability to toggle opening a number of window views?

None of these other sandbox links work? https://prod.imodeljs.org/reference/core-frontend/views/

Tony

**Comment by NancyMcCallB** [comment]
@VedankNaik and @tonygem2022 ,

You can change the ContentGroup for the active view with something like this:

`          const contentGroup = new ContentGroup(contentGroupProps); // your defined ContentGroupProps
          await FrontstageManager.setActiveContentGroup(contentGroup);
`

Please let me know if you run into any trouble with it.

Thanks,
Nancy

**Reply by VedankNaik** [reply]
Thanks. This worked
